clients/src/main/java/org/apache/kafka/clients/consumer/internals/CoordinatorManager.java

```diff
@@ -0,0 +1,245 @@
+    private final Optional<String> groupId;
+
+    public CoordinatorManager(final Time time,
+                              final LogContext logContext,
+                              final ConsumerConfig config,
+                              final BlockingQueue<BackgroundEvent> backgroundEventQueue,
+                              final Optional<String> groupId,
+                              final long rebalanceTimeoutMs) {
+        this.time = time;
+        this.log = logContext.logger(this.getClass());
+        this.backgroundEventQueue = backgroundEventQueue;
+        this.exponentialBackoff = new ExponentialBackoff(
+                config.getLong(ConsumerConfig.RECONNECT_BACKOFF_MS_CONFIG),
+                RECONNECT_BACKOFF_EXP_BASE,
+                config.getLong(CommonClientConfigs.RECONNECT_BACKOFF_MAX_MS_CONFIG),
+                RECONNECT_BACKOFF_JITTER);
+        this.coordinatorRequestState = new CoordinatorRequestState();
+        this.groupId = groupId;
+        this.rebalanceTimeoutMs = rebalanceTimeoutMs;
+        this.requestTimeoutMs = config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+    }
+
+    /**
+     * Returns a non-empty UnsentRequest we need send a FindCoordinatorRequest. These conditions are:
+     * 1. The request has not been sent
+     * 2. If the previous request failed, and the retryBackoff has expired
+     * @return Optional UnsentRequest.  Empty if we are not allowed to send a request.
+     */
+    public Optional<NetworkClientDelegate.UnsentRequest> tryFindCoordinator() {
+        if (coordinatorRequestState.lastSentMs == -1) {
+            // no request has been sent
+            return Optional.of(
+                    new NetworkClientDelegate.UnsentRequest(
+                            this.time.timer(requestTimeoutMs),
+                            getFindCoordinatorRequest(),
+                            new FindCoordinatorRequestHandler()));
+        }
+
+        if (coordinatorRequestState.lastReceivedMs == -1 ||
+                coordinatorRequestState.lastReceivedMs < coordinatorRequestState.lastSentMs) {
+            // there is an inflight request
+            return Optional.empty();
+        }
+
+        if (!coordinatorRequestState.requestBackoffExpired()) {
+            // retryBackoff
+            return Optional.empty();
+        }
+
+        return Optional.of(
+                new NetworkClientDelegate.UnsentRequest(
+                        this.time.timer(requestTimeoutMs),
+                        getFindCoordinatorRequest(),
+                        new FindCoordinatorRequestHandler()));
+    }
+
+    /**
+     * Mark the current coordinator null and return the old coordinator. Return an empty Optional
+     * if the current coordinator is unknown.
+     * @param cause why the coordinator is marked unknown
+     * @return Optional coordinator node that can be null.
+     */
+    protected Optional<Node> markCoordinatorUnknown(String cause) {
```

Comment:
What is the use case where we want the old coordinator?

Reply:
I'm not sure, maybe we could just make it a void type. (the signature is from the old impl)

Reply:
In the background thread - we will need to check if the coordinator and disconnect the old one.

Reply:
Is it necessary to proactively disconnect the old coordinator?

Reply:
Could you elaborate a bit here? Do you mean we could rely on networkclient poll to handle the disconnect?